File: mir-service/src/main/kotlin/it/unibo/harmonikt/handlers/RobotHandlers.kt
package it.unibo.harmonikt.handlers

import io.ktor.http.HttpStatusCode
import io.ktor.server.request.receive
import io.ktor.server.resources.delete
import io.ktor.server.resources.get
import io.ktor.server.resources.post
import io.ktor.server.response.respond
import io.ktor.server.routing.Routing
import it.unibo.harmonikt.api.dto.RobotIdDTO
import it.unibo.harmonikt.api.dto.RobotRegistrationDTO.MirRobotRegistrationDTO
import it.unibo.harmonikt.api.dto.RobotStatusDTO
import it.unibo.harmonikt.model.Action
import it.unibo.harmonikt.model.Robot
import it.unibo.harmonikt.model.RobotType
import it.unibo.harmonikt.repository.MirRobotRepository
import it.unibo.harmonikt.resources.MirRobots

/**
 * Handles HTTP requests related to robots in the MIR service.
 */
object RobotHandlers {
    /**
     * Sets up the routing for robot-related endpoints.
     *
     * @param repository The MirRobotRepository instance to handle robot data.
     */
    fun Routing.setupRobotHandlers(repository: MirRobotRepository) {
        // GET /robots - Retrieve all active robots
        get<MirRobots> {
            val robots = repository.getRobots()
            call.respond(robots)
        }

        // POST /robots - Add a new robot
        post<MirRobots> {
            val request = call.receive<MirRobotRegistrationDTO>()
            val robotId = repository.createRobot(
                request.canonicalName,
                request.apiToken,
                request.host,
            )
            if (robotId != null) {
                call.respond(HttpStatusCode.Created, RobotIdDTO(robotId))
            } else {
                call.respond(HttpStatusCode.Conflict, "Robot ${request.canonicalName} not created.")
            }
        }

        // DELETE /robots/{robotId} - Remove a robot
        delete<MirRobots.Id> { robot ->
            val done = repository.deleteRobot(robot.id)
            if (done) {
                call.respond(HttpStatusCode.OK, "Robot with id ${robot.id} deleted")
            } else {
                call.respond(HttpStatusCode.NotFound, "Delete failed for robot with id ${robot.id}, possibly not found")
            }
        }

        // GET /robots/{robotId} - Retrieve information about a robot
        get<MirRobots.Id> { r ->
            val robot: Robot? = repository.getRobotById(r.id)
            if (robot != null) {
                val dto = RobotStatusDTO(
                    id = robot.id,
                    canonicalName = robot.name,
                    state = robot.currentState,
                    batteryLevel = robot.batteryLevel,
                    currentPosition = robot.currentPosition,
                    type = RobotType.MIR,
                )
                call.respond(dto)
            } else {
                call.respond(HttpStatusCode.NotFound, "Robot with id ${r.id} not found")
            }
        }

        // POST /robots/{robotId}/actions - Send an action to a robot
        post<MirRobots.Id.Move> { action ->

        }
    }
}

File: spot-service/app/handlers/robot_handlers.py
"""
Robot Handlers Module

This module defines the handlers for robot-related HTTP endpoints in the Spot service.
These handlers process HTTP requests and generate appropriate responses for robot operations.
"""

from fastapi import APIRouter, HTTPException, status
from typing import List, Dict
from uuid import UUID
from pydantic import BaseModel
from typing import Union

from app.repositories.spot_robot_repository import RobotRepository
from app.api.robot_api import (
    RobotRegistrationDTO,
    RobotInfoDTO,
    RobotStatusDTO,
    RobotIdDTO,
)
from app.services.spot_robot_service import RobotService

# Create a router for robot endpoints
router = APIRouter()


def setup_robot_handlers(repository: RobotRepository, service: RobotService):
    """
    Sets up the routing for robot-related endpoints in the Spot service.

    Args:
        repository: The repository for accessing robot data
        service: The service for handling robot actions

    Returns:
        The configured router
    """

    @router.get("/robots", response_model=List[RobotInfoDTO])
    async def get_robots():
        """Get all robot IDs"""
        return [RobotInfoDTO.from_robot(robot) for robot in repository.get_robots()]

    @router.post("/robots", response_model=RobotIdDTO)
    async def create_robot(robot_creation_request: RobotRegistrationDTO):
        """Create a new robot"""
        try:
            robot_id = repository.create_robot(
                username=robot_creation_request.username,
                password=robot_creation_request.password,
                address=robot_creation_request.host,
                canonical_name=robot_creation_request.canonicalName,
            )
            return RobotIdDTO(id=robot_id)
        except ValueError as e:
            raise HTTPException(status_code=400, detail=str(e))

    @router.get("/robots/{robot_id}", response_model=RobotStatusDTO)
    async def get_robot(robot_id: UUID):
        """Get a robot by ID"""
        robot = repository.get_robot_by_id(robot_id)
        if robot is None:
            raise HTTPException(
                status_code=404, detail=f"Robot with ID {robot_id} not found"
            )
        return RobotStatusDTO.from_robot(robot)

    @router.delete("/robots/{robot_id}", response_model=Dict[str, str])
    async def delete_robot(robot_id: UUID):
        """Delete a robot by ID"""
        if not repository.delete_robot(robot_id):
            raise HTTPException(
                status_code=404, detail=f"Robot with ID {robot_id} not found"
            )
        return {"message": f"Robot with ID {robot_id} deleted successfully"}

    @router.post("/robots/{robot_id}/actions", status_code=status.HTTP_201_CREATED)
    async def create_marker(robot_id: UUID, action: ActionDTO):
        """Create a new action for a robot"""
        try:
            if isinstance(action, MoveToTargetDTO):
                service.move_to_target(
                    robot_id=robot_id,
                    fiducial=action.fiducial,
                )
            else:
                raise HTTPException(status_code=400, detail="Unsupported action type")
            return {"message": f"Creation for robot {robot_id}"}
        except ValueError as e:
            raise HTTPException(status_code=400, detail=str(e))

    return router


type ActionDTO = Union[MoveToTargetDTO]


class MoveToTargetDTO(BaseModel):
    type: str = "MoveToTarget"
    fiducial: int
